// astro/dist/cli/install-package.js
import { createRequire } from "node:module";
import { sep } from "node:path";
import { pathToFileURL } from "node:url";
import boxen from "boxen";
import { execa } from "execa";
import { bold, cyan, dim, magenta } from "kleur/colors";
import ora from "ora";
import prompts from "prompts";
import resolvePackage from "resolve";
import whichPm from "which-pm";
import {} from "../core/logger/core.js";
const require2 = createRequire(import.meta.url);
async function getPackage(packageName, logger, options, otherDeps = []) {
  try {
    if (packageName === "@astrojs/db") {
      const packageJsonLoc = require2.resolve(packageName + "/package.json", {
        paths: [options.cwd ?? process.cwd()]
      });
      const packageLoc = pathToFileURL(packageJsonLoc.replace(`package.json`, "dist/index.js"));
      const packageImport2 = await import(packageLoc.toString());
      return packageImport2;
    }
    await tryResolve(packageName, options.cwd ?? process.cwd());
    const packageImport = await import(packageName);
    return packageImport;
  } catch (e) {
    if (options.optional)
      return void 0;
    logger.info(
      "SKIP_FORMAT",
      `To continue, Astro requires the following dependency to be installed: ${bold(packageName)}.`
    );
    const result = await installPackage([packageName, ...otherDeps], options, logger);
    if (result) {
      const packageImport = await import(packageName);
      return packageImport;
    } else {
      return void 0;
    }
  }
}
function tryResolve(packageName, cwd) {
  return new Promise((resolve, reject) => {
    resolvePackage(
      packageName,
      {
        basedir: cwd
      },
      (err) => {
        if (err) {
          reject(err);
        } else {
          resolve(0);
        }
      }
    );
  });
}
function getInstallCommand(packages, packageManager) {
  switch (packageManager) {
    case "npm":
      return { pm: "npm", command: "install", flags: [], dependencies: packages };
    case "yarn":
      return { pm: "yarn", command: "add", flags: [], dependencies: packages };
    case "pnpm":
      return { pm: "pnpm", command: "add", flags: [], dependencies: packages };
    case "bun":
      return { pm: "bun", command: "add", flags: [], dependencies: packages };
    default:
      return null;
  }
}
async function installPackage(packageNames, options, logger) {
  const cwd = options.cwd ?? process.cwd();
  const packageManager = (await whichPm(cwd)).name ?? "npm";
  const installCommand = getInstallCommand(packageNames, packageManager);
  if (!installCommand) {
    return false;
  }
  const coloredOutput = `${bold(installCommand.pm)} ${installCommand.command}${[
    "",
    ...installCommand.flags
  ].join(" ")} ${cyan(installCommand.dependencies.join(" "))}`;
  const message = `
${boxen(coloredOutput, {
    margin: 0.5,
    padding: 0.5,
    borderStyle: "round"
  })}
`;
  logger.info(
    "SKIP_FORMAT",
    `
  ${magenta("Astro will run the following command:")}
  ${dim(
      "If you skip this step, you can always run it yourself later"
    )}
${message}`
  );
  let response;
  if (options.skipAsk) {
    response = true;
  } else {
    response = (await prompts({
      type: "confirm",
      name: "askToContinue",
      message: "Continue?",
      initial: true
    })).askToContinue;
  }
  if (Boolean(response)) {
    const spinner = ora("Installing dependencies...").start();
    try {
      await execa(
        installCommand.pm,
        [installCommand.command, ...installCommand.flags, ...installCommand.dependencies],
        { cwd }
      );
      spinner.succeed();
      return true;
    } catch (err) {
      logger.debug("add", "Error installing dependencies", err);
      spinner.fail();
      return false;
    }
  } else {
    return false;
  }
}
export {
  getPackage
};
